cli/state.go

```diff
@@ -1868,3 +1826,144 @@ var stateNtwkVersionCmd = &cli.Command{
 		return nil
 	},
 }
+
+var stateDealCmd = &cli.Command{
+	Name:  "deal",
+	Usage: "Get information about deal(s)",
+	Subcommands: []*cli.Command{
+		stateGetDealCmd,
+		stateDealFeesCmd,
+	},
+}
+
+var stateGetDealCmd = &cli.Command{
+	Name:      "get",
+	Usage:     "View on-chain deal info",
+	ArgsUsage: "[dealId]",
+	Action: func(cctx *cli.Context) error {
+		api, closer, err := GetFullNodeAPI(cctx)
+		if err != nil {
+			return err
+		}
+		defer closer()
+
+		ctx := ReqContext(cctx)
+
+		if !cctx.Args().Present() {
+			return fmt.Errorf("must specify deal ID")
+		}
+
+		dealid, err := strconv.ParseUint(cctx.Args().First(), 10, 64)
+		if err != nil {
+			return xerrors.Errorf("parsing deal ID: %w", err)
+		}
+
+		ts, err := LoadTipSet(ctx, cctx, api)
+		if err != nil {
+			return err
+		}
+
+		deal, err := api.StateMarketStorageDeal(ctx, abi.DealID(dealid), ts.Key())
+		if err != nil {
+			return err
+		}
+
+		data, err := json.MarshalIndent(deal, "", "  ")
+		if err != nil {
+			return err
+		}
+		fmt.Println(string(data))
+
+		return nil
+	},
+}
+
+var stateDealFeesCmd = &cli.Command{
+	Name:  "fees",
+	Usage: "View the storage fees associated with a particular deal or storage provider",
+	Flags: []cli.Flag{
+		&cli.StringFlag{
+			Name:  "provider",
+			Usage: "provider whose outstanding fees you'd like to calculate",
+		},
+		&cli.IntFlag{
+			Name:  "dealId",
+			Usage: "deal whose outstanding fees you'd like to calculate",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		api, closer, err := GetFullNodeAPI(cctx)
+		if err != nil {
+			return err
+		}
+		defer closer()
+
+		ctx := ReqContext(cctx)
+
+		ts, err := LoadTipSet(ctx, cctx, api)
+		if err != nil {
+			return err
+		}
+
+		var ht abi.ChainEpoch
+		if ts == nil {
+			head, err := api.ChainHead(ctx)
+			if err != nil {
+				return err
+			}
+
+			ht = head.Height()
+		} else {
+			ht = ts.Height()
+		}
+
+		if cctx.IsSet("provider") {
+			p, err := address.NewFromString(cctx.String("provider"))
+			if err != nil {
+				return fmt.Errorf("failed to parse provider: %w", err)
+			}
+
+			deals, err := api.StateMarketDeals(ctx, ts.Key())
```

Comment:
I'd avoid this API call, it will return all the milions of deals in the market actor, calling it with curl the response is 366M and takes 1.5min

Reply:
@arajasek thoughts on this?